server/mqtt.go

```diff
@@ -2199,25 +2232,6 @@ func (as *mqttAccountSessionManager) processSubs(sess *mqttSession, c *client,
 		sess.cons[sid] = cc
 	}
 
-	// Helper that sets the sub's mqtt fields and possibly serialize retained messages.
-	// Assumes account manager and session lock held.
-	setupSub := func(sub *subscription, qos byte) {
-		subs := []*subscription{sub}
-		if len(sub.shadow) > 0 {
-			subs = append(subs, sub.shadow...)
-		}
-		for _, sub := range subs {
```

Comment:
I am not too sure of your change. We used to create a `&mqtt{}` per shadow sub, now you rely on processSub() to have the shadow be a copy of the sub, but I am pretty sure that this does not make a deep copy of the mqtt{} struct. Which means that things in mqtt{} will be shared between the original sub and its shadow subscriptions. I honestly don't know if this is an issue or not, but given the `prm` which is a writer for serializing retained messages, I feel like sharing this is wrong.

Reply:
Right, I can debate how to fix this, but your other point I think leaves this one moot indeed if the entire session needs to be locked, and I think you are right, it would be.

Going to look some more into doing everything JS in advance of `c.processSub`. Need to get the shadow subjects without actually effecting the sub.

Reply:
Or maybe for the serialization of retained messages, we then go to more fine grained locking of the session, which may mean simply releasing/reacquiring the lock before loading the message (which I usually don't like to do, but, something to think about).

Reply:
I am wary of either, TBH, because I am not sure how to interpret `Any existing retained messages matching the Topic Filter MUST be re-sent, but the flow of publications MUST NOT be interrupted [MQTT-3.8.4-3].`

One interpretation would be that we could even deliver them in a separate go-routine if we so desired, as long as new incoming PUBLISH delivered as soon as we send out SUBACK. Another interpretation would be that the _order_ must not be interrupted, so no new messages should be delivered until all retained (that are updated close to real-time) are sent out. The intention of the current code was the latter.

Anyway, looking into doing all necessary JS before `processSub`. Besides retained messages, we also need to create the PubRel consumer/subscription (1-time/sess).

Reply:
Yes, and thank you for reminded me of another reason why locking was held for that whole duration. Exactly what you said about interpretation of delivery of messages.